src/items/functions.md

```diff
@@ -189,6 +189,50 @@ Exhaustive list of permitted structures in const functions:
   the following unsafe operations:
     * calls to const unsafe functions
 
+## Async functions
+
+Functions may be declared as async, and this can also be combined with the
+`unsafe` quantifier:
+
+```rust
+async fn regular_example() { }
+async unsafe fn unsafe_example() { }
+```
+
+Async functions do not actually do any work when called: instead, they
+capture their arguments into a future. That future, when polled, will
+execute the function body.
+
+Roughly speaking, an async function can be desugared into a function
+that returns `impl Future` and with an `async` block as its body:
+
+```rust
+// Source
+async fn example(x: &str) -> usize {
+  x.len()
+}
+
+// Result
+fn example<'a>(x: &'a str) -> impl Future<Output = usize> + 'a {
+  async move { x.len() }
+}
+```
+
+The actual desugaring is performed internally to the compiler and is mildly
+more complex:
+
+- The `impl Future` which returns is assumed to capture all lifetime
+  parameters from the `async fn` declaration. This can be seen in the
+  desugared example above, which explicitly outlives (and hence
+  captures) `'a`.
+- The `async move` in the body includes some code to capture otherwise
+  unused function parameters. This code ensures that function
+  parameters are dropped in the same order as they would be if the
```

Comment:
A link to drop order would be good; However, @matthewjasper's PR hasn't landed so please make a follow-up issue.

Reply:
I've added a note to #634